File: octopus.cpp
// Solution for Octopus (HSPT 2021 - On Site)
// Author: Atharva Nagarkar

#include <bits/stdc++.h>

using namespace std;

/*
* For this problem, one can see that performing a right cyclic shift on a subarray
* is the same as taking the last element in the subarray and placing it in front of
* the subarray. This can be generalized to simply taking one element out of the array
* and placing it somewhere else. Now we wish to see the minimum number elements to
* move to make the array sorted. This is equivalant to keeping maximum number of things
* the same in the array; one can realize that this value is the LIS (longest increasing subsequence)
* in the array. Thus, the answer to our problem is N - LIS(array), the total number of elements
* minus maximum number to keep the same = minimum number to move.
*/

int main() {

    // read in number of test cases
    int tests;
    cin >> tests;

    while(tests--) {

        // read in number of integers in the array
        int n;
        cin >> n;
        
        // keep a map with an invariant, which is that if there are two keys
        // in the map x < y, then dp[x] < dp[y]. If this is not the case,
        // then keeping around y is useless since you can build off x with a
        // strictly better answer
        map<int, int> dp;

        int ans = 1;
        for(int i = 0; i < n; ++i) {
            int x;
            cin >> x;

            // first lets see where to insert this element
            auto it = dp.upper_bound(x);

            // if this is the first element in the array, then it's answer is just 1
            if(it == dp.begin()) {
                dp[x] = 1;
                continue;
            }

            // get the best answer of the previous element
            // our current answer is the previous element's answer + 1
            it = prev(it);
            dp[x] = it->second + 1;
            ans = max(ans, dp[x]);

            // after inserting the answer for the current element, a bunch of entries
            // so while bad entries exist, remove them
            while(true) {
                it = dp.upper_bound(x);
                if(it == dp.end()) break;
                if(it->second <= dp[x]) {
                    dp.erase(it);
                } else {
                    break;
                }
            }
        }

        // print the answer!
        cout << n - ans << '\n';

    }

    return 0;
}

//69

File: octopus.java
// 2021 UCF HSPT - Octopus Garden
// Author: Natalie Longtin

/*
observe that performing a cyclic shift in either direction is equivalent
to taking a single element of the array and placing it somewhere else.
our problem is then reduced to: what is the minimum number of elements
we need to move somewhere else in order to sort the array?

since it never makes sense to move the same element more than once, we just need to
choose which elements to move and which elements to keep in the same spot.
we want to leave as many in the same place as possible.

observe that this maximum value is equal to the maximum number
that are already in the correct relative order,
subtracted from the total number of elements in the array.
and so our problem is further reduced to finding the longest increasing subsequence (LIS)
of the array; the answer will be n - LIS.
*/

import java.util.Arrays;
import java.util.Scanner;

public class octopus {
	public static void main(String[] args) {
		Scanner scan=new Scanner(System.in);
		
		//process test cases
		int t=scan.nextInt();
		for(int tt=0;tt<t;tt++) {
			int n=scan.nextInt();
			int[] a=new int[n];
			for(int i=0;i<n;i++) a[i]=scan.nextInt();
			
			//lis(a) elements will already be in the correct order
			//this means that n-lis(a) elements are out of place and must be moved
			System.out.println(n-lis(a));
		}
	}
	
	//nlog(n) implementation of longest increasing subsequence
	//the principle is that we add the elements to the LIS one by one,
	//overwriting the earliest element we can each time,
	//and keeping track of the longest LIS we have encountered so far.
	
	public static int lis(int[] a) {
		int n=a.length;
		int inf=Integer.MAX_VALUE/2-5;
		int[] insert=new int[n+1];
		
		Arrays.fill(insert,inf);
		insert[0]=-inf;
		int[] lis=new int[n];
		int res=0;
		
		for(int i=0;i<n;i++) {
			int idx=search(insert,a[i]);
			insert[idx]=a[i];
			lis[i]=idx;
			res=Math.max(res,lis[i]);
		}
		return res;
	}
	
	//search(a,x) returns the smallest index such that a[i]>=x
	public static int search(int[] a, int x) {
		int res=-1;
		
		int n=a.length;
		int lo=0, hi=n-1;
		while(hi>=lo) {
			int mid=(lo+hi)/2;
			if(a[mid]>=x) {
				res=mid;
				hi=mid-1;
			}
			else lo=mid+1;
		}
		return res;
	}
}
